numpy/core/src/multiarray/refcount.c

```diff
@@ -444,7 +444,12 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
             optr += inner_elsize;
         }
     }
-    else {
+    else if (PyDataType_FLAGCHK(dtype, NPY_ITEM_IS_POINTER))
+    {
```

Comment:
```suggestion
    else if (PyDataType_FLAGCHK(dtype, NPY_ITEM_IS_POINTER)) {
```
Super nit on the bracket.

Should we just check for not legacy, maybe?  The assert branch is pretty unnecessary.  (My reason is that I am not sure the flag check isn't always true in practice, at which point it feels a bit useless to even keep the assert.)

Then add a comment, that this function should be used for `empty()` to fill with `None` and we assume that non-legacy DTypes are OK with `NULL` initialization in this case (for now).

(N.B. this function looks terribly slow for structured dtypes.)


The rest looks perfect, thanks.

Reply:
I think assuming that all non-legacy dtypes that have `NPY_ITEM_REFCOUNT` set actually represent heap-allocated types should be fine for now, with the comment. I agree this is something we need to come back to, especially if there ends up being more places inside numpy where the logic needs clarifying.

I'll leave the `assert(0)` just in case it can somehow be hit in the future.

Reply:
> actually represent heap-allocated types

Or I guess are fine with being initialized to NULL, with the dtype ultimately being responsible for managing reference counts in the clear function.

Reply:
Hmmmpf, this is public API and it actually already does something for non-referenced versions (calls setitem directly _uggggh_).  This is a mess, but it might be good to go into the non-referenced path here.

Reply:
:+1: see [60ad87d](https://github.com/numpy/numpy/pull/23262/commits/60ad87d860d21b62412dc10e64f78e79ad886fe0)